proposals/306-ipv6-happy-eyeballs.txt

```diff
@@ -48,135 +48,103 @@ Ticket: https://trac.torproject.org/projects/tor/ticket/29801
-   structure, we need to fill in both the IPv4 and IPv6 address if they both
-   are available. If only one family is available for a relay (IPv4 or IPv6),
-   we should fill in the address for preferred family and leave the alternate
-   family null.
+     * Bootstrapping clients can connect to fallback directory mirrors or
+       authorities
 
-2.2. Relay Selection Changes
-
-   In Proposal 283, we have allowed microdescriptor consensus documents to
-   contain IPv6 addresses. As clients download microdescriptors, Prop283
-   makes it possible to implement Prop306.
-
-   When we select candidates for the entry guard, we should select at least
-   one relay with an IPv6 address. This makes it possible for an IPv6-only
-   client to bootstrap. Otherwise we would have failures if all the selected
-   guard candidates only have IPv4 addresses.
-
-3. Relay Connections
+     * v3 single onion services can use IPv4 or IPv6 addresses to connect
+       to introduction and rendezvous points, and
 
-   If there is an existing authenticated connection, we should use it
-   similar to how we used it pre-Prop306.
+     * Clients can be configured to disable entry guards
 
-   If there is no existing authenticated connection for an extend info,
-   we should attempt to connect using the first available, allowed, and
-   preferred address.
+   Bridges are out of scope for this proposal, because Tor does not support
+   multiple IP addresses in a single bridge line.
 
-   We should also allow falling back to the alternate address. For this,
-   three alternate designs will be given.
-
-3.1. Proposed Designs
-
-   This subsection will have three proposed designs for connecting to relays
-   via IPv4 and IPv6 in a Tor implementation of Happy Eyeballs.
-
-   These proposed designs will have some tradeoffs, including:
-
-    * Launching multiple TCP connections places up to 2x extra socket load on
-      dual-stack relays and authorities, because both connections may succeed.
-
-    * Launching multiple TLS connections places up to 2x the CPU load on
-      dual-stack relays and authorities, because both connections may succeed.
-
-    * Increasing the delay between connections mitigates the above issues,
-      but reduces perceived performance, particularly at bootstrap time
-      (pre-emptive circuits hide these delays after bootstrap).
+   The extend info structure should contain both an IPv4 and an IPv6 address.
+   This will allow us to try IPv4 and the IPv6 addresses should both be
+   available on a relay and the client is dual-stack.
 
-   The proposed designs are as listed as follows:
+   When parsing relay descriptors and filling in the extend info data
+   structure, we need to fill in both the IPv4 and IPv6 address if both are
+   available. If only one family is available for a relay (IPv4 or IPv6), we
+   should fill in the address for preferred family and leave the alternate
+   family null.
 
```

Comment:
I think we're missing a section here about bootstrapping changes. I'll commit all my suggested changes, then try writing it up.

Reply:
Added in 2bb59d4.